### PR TITLE
NXDOC-1337: Modify metalsmith build to handle doc assets urls

### PR DESCRIPTION
**Pull request by brunoribeiro**
created metalsmith plugin to handle special urls that must be downloaded from nuxeo server.
Please review as my js knowledge is limited.

### CODE REVIEW

File: modules/metalsmith/doc_assets.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+        }
+
+        Promise.all(file_promises)
+            .then(() => {
+                return done();
+            })
+            .catch((err) => {
+                error(err);
+                return done();
```
Comment (woodyrew) [inline modules/metalsmith/doc_assets.js]: also pass `err` to `done` > `return done(err);`

File: modules/metalsmith/doc_assets.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            auth   : {
+                method  : 'basic',
+                username: process.env.NX_ASSETS_USER || 'Administrator',
+                password: process.env.NX_ASSETS_PWD || 'Administrator'
+            },
+        });
+
+        const file_promises = [];
+
+        for (const filename in files) {
```
Comment (woodyrew) [inline modules/metalsmith/doc_assets.js]: My preference would be a functional style and move the `selector` definition outside of the loop.
```js
const selector = Object.keys(map).join();
Object.keys(files).forEach((filename) => {
  if (is_html(filename)) {
    file_promises.push(check_file(files, filename, selector, nuxeo));
  }
})
```
It's not wrong, just style preference.

File: modules/metalsmith/doc_assets.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
+'use strict';
+/* eslint-env es6 */
+
+
+const {info, error} = require('./../debugger')('nx_assets');
```
Comment (woodyrew) [inline modules/metalsmith/doc_assets.js]: just `'../debugger'` is ok.

File: modules/metalsmith/doc_assets.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    // 'link'  : 'href',
+    // 'script': 'src',
+    'img'  : 'src',
+    'video': 'src',
+    'audio': 'src'
+    // 'source': 'src'
+};
+
+
+const checksum = function (str, algorithm, encoding) {
```
Comment (nelsonsilva) [inline modules/metalsmith/doc_assets.js]: Perhaps we could make use of ES2015 default parameters and arrow functions, something like:
```const checksum = (str, algorithm = 'md5', encoding = 'utf8') => crypto....```

Comment (nelsonsilva) [reply modules/metalsmith/doc_assets.js]: Same thing for other functions below...
`const is_html = (file) => /\.html?/.test(file);`

File: modules/metalsmith/doc_assets.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    return new Promise(function (resolve) {
+
+        const file = files[filename];
+
+        const contents = file.contents.toString();
+        const $ = cheerio.load(contents);
+
+        const url_promises = [];
+
+        $(selector).each(function (i, el) {
```
Comment (nelsonsilva) [inline modules/metalsmith/doc_assets.js]: I'd probably factor this into functions to have something like:
`Promise.all(
   $(selector)
    .filter(_nxAssetUrls)
    .map(_fetchAsset)
)`
where `_fetchAsset` would return a Promise

File: modules/metalsmith/doc_assets.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                        if (!doc.properties['file:content']) {
+                            error('nx_asset without file:content');
+                            return null;
+                        }
+
+                        // build the right url
+                        const folder = 'nx_assets';
+                        const ext = /(?:\.([^.]+))?$/.exec(doc.properties['file:content'].name)[1];
+                        const asset_type = doc.properties['doc_asset:nature'];
+                        const asset_file = folder + '/' + uid + '+' + asset_type + '.' + ext;
```
Comment (nelsonsilva) [inline modules/metalsmith/doc_assets.js]: You can use template literals like
```
`${folder}/${uid}+${asset_type}.${ext}`
```

File: modules/metalsmith/doc_assets.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                        const href = '/' + asset_file;
+
+                        $el.attr(attr, href);
+                        const html = $.html();
+                        file.contents = new Buffer(html);
+                        files[filename] = file;
+
+                        // if file already exist in fs
+                        if (fs.existsSync(asset_file)) {
```
Comment (nelsonsilva) [inline modules/metalsmith/doc_assets.js]: This should be using the async version and resolving the promise with the file to make it transparent to the processing logic

File: modules/metalsmith/doc_assets.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                            });
+                    })
+                    .catch(function (err) {
+                        throw err;
+                    });
+
+                url_promises.push(p);
+            }
+
+            Promise.all(url_promises)
```
Comment (nelsonsilva) [inline modules/metalsmith/doc_assets.js]: No need to create the wrapper promise in the beginning to call `resolve()` here. You can just `return Promise.all([...])`

File: modules/metalsmith/doc_assets.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+    });
+};
+
+
+const doc_assets = function () {
+
+    return function (files, metalsmith, done) {
+
+        const nuxeo = new Nuxeo({
```
Comment (nelsonsilva) [inline modules/metalsmith/doc_assets.js]: This could be moved out of the function I guess

File: modules/metalsmith/doc_assets.js [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            if (!is_html(filename)) {
+                continue;
+            }
+
+            const selector = Object.keys(map).join();
+            file_promises.push(check_file(files, filename, selector, nuxeo));
+
+        }
+
+        Promise.all(file_promises)
```
Comment (nelsonsilva) [inline modules/metalsmith/doc_assets.js]: Again just
```
return Promise.all(
  Object.keys(map)
    .filter(isHtml)
    .map((filename)=> check_file(..))
```

File: nx_assets/readme.txt [hunk 11]
```diff
@@ -0,0 +1 @@
+nx_assets got here.
```
Comment (nelsonsilva) [inline nx_assets/readme.txt]: If goal is to keep this folder just `touch .gitignore`

Comment (woodyrew) [reply nx_assets/readme.txt]: If you don't want to commit any of the files use the following in
`nx_assets/.gitignore`:
```
# Ignore everything in this directory
*
# Except this file
!.gitignore
```

**Pull request merged**